testplan/common/utils/process.py

```diff
@@ -84,6 +87,57 @@ def kill_process(proc, timeout=5, signal_=None, output=None):
     return proc.returncode
 
 
+def kill_process_ex(proc, timeout=5, signal_=None, output=None):
+    """
+    If alive, kills the process (an instance of ``psutil.Process``).
+    Try killing the child process at first and then killing itself.
+    First call ``terminate()`` or pass ``signal_`` if specified
+    to terminate for up to time specified in timeout parameter.
+    If process hangs then call ``kill()``.
+
+    :param proc: process to kill
+    :type proc: ``psutil.Process``
+    :param timeout: timeout in seconds, defaults to 5 seconds
+    :type timeout: ``int``
+    :param output: Optional file like object for writing logs.
+    :type output: ``file``
+    :return: List of processes which are still alive
+    :rtype: ``list`` or ``psutil.Process``
+    """
+    _log = functools.partial(_log_proc, output=output)
+
+    procs = proc.children(recursive=True) + [proc]
```

Comment:
I know that this is the same logic as used in `kill_process`. It doesn't need changing for this PR, but in future I think we should change the high-level logic so that it does:

1. Send SIGTERM to parent process __only__. Normally, the parent process should handle terminating its children.
2. Wait for all processes to exit.
3. Kill any remaining processes with SIGKILL

Probably we could share some of the waiting / kill remaining logic with `kill_process`.

Reply:
Not sure how the parent process deal with child, for CouchDB, if the parent process receives a TERM signal, then all of its children will also be terminated, that sounds reasonable. So I agree with this logic, if a parent process does not handle its children on receiving TERM or INT, it means the code is not correct, then we can arbitrarily kill all of its children. I'll update the PR in this way.

Reply:
Yes exactly, a well-behaved process should terminate its children before it exits, we just need to kill them as a backup. On the other hand, a process may not expect its children to terminate  and may e.g. try to restart them if it detects that, which maybe could cause problems.